src/coreclr/System.Private.CoreLib/src/System/Reflection/MdImport.cs

```diff
@@ -304,9 +304,19 @@ internal static unsafe MarshalAsAttribute GetMarshalAs(ConstArray nativeType, Ru
         #endregion
 
         #region Constructor
-        internal MetadataImport(IntPtr metadataImport2, object? keepalive)
+        public MetadataImport()
         {
-            m_metadataImport2 = metadataImport2;
+            m_metadataImport2 = IntPtr.Zero;
+            m_keepalive = null;
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern unsafe IntPtr GetMetadataImport(RuntimeModule module);
+
+        internal MetadataImport(RuntimeModule module, object keepalive)
+        {
+            ArgumentNullException.ThrowIfNull(module);
+            m_metadataImport2 = GetMetadataImport(module);
             m_keepalive = keepalive;
```

Comment:
The handling of keepalive is questionable. The JIT can inline this constructor, observe that `m_keepalive` is unused, and optimize it out so it won't serve the intended purpose.

We may be lucky today that it is not causing any observable crash. Harmless refactorings like this one may turn this into observable crash.

We may want to refactor this into something more robust. For example:
- Delete keepalive here
- Add IntPtr m_metadataImport2 as a field to `RuntimeModule`. It would address my other comment about potential perf impact.
- Make sure that all places that use m_metadataImport2 keep the RuntimeModule alive for as long as it is used (e.g. add GC.KeepAlive at the end of the method, we have many places like that)